From df2de4ff22bb1a0adc7826d5520713f2f3450024 Mon Sep 17 00:00:00 2001
From: Zhou Wang <wangzhou1@hisilicon.com>
Date: Fri, 23 May 2025 20:45:29 +0800
Subject: [PATCH 599/960] downstream: irqchip/gic: Add HiSilicon PV SGI support

Use the smccc interface to enable ipiv for the guest OS, modify the
guest kernel, and use multiple unicasts to implement group boradcast.
In this way, do not need to modify the MPIDR.

In addition, the MPIDR modification is deleted, and the GICD
configuration is modified. The hardware uses the mpidr to calculate
the corresponding vCPU ID to lookup vpeid table.

Signed-off-by: Zhou Wang <wangzhou1@hisilicon.com>
Signed-off-by: Jinqian Yang <yangjinqian1@huawei.com>
Signed-off-by: nzq-hw <nizhiqiang1@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/kvm/sys_regs.c    | 14 ------------
 drivers/irqchip/irq-gic-v3.c | 42 ++++++++++++++++++++++++++++++++++--
 2 files changed, 40 insertions(+), 16 deletions(-)

diff --git a/arch/arm64/kvm/sys_regs.c b/arch/arm64/kvm/sys_regs.c
index f6e19275c2aa..72f83f246323 100644
--- a/arch/arm64/kvm/sys_regs.c
+++ b/arch/arm64/kvm/sys_regs.c
@@ -694,10 +694,6 @@ static u64 reset_actlr(struct kvm_vcpu *vcpu, const struct sys_reg_desc *r)
 	return actlr;
 }
 
-#ifdef CONFIG_ARM64_HISI_IPIV
-extern struct static_key_false ipiv_enable;
-#endif
-
 static u64 reset_mpidr(struct kvm_vcpu *vcpu, const struct sys_reg_desc *r)
 {
 	u64 mpidr;
@@ -713,16 +709,6 @@ static u64 reset_mpidr(struct kvm_vcpu *vcpu, const struct sys_reg_desc *r)
 	mpidr |= ((vcpu->vcpu_id >> 4) & 0xff) << MPIDR_LEVEL_SHIFT(1);
 	mpidr |= ((vcpu->vcpu_id >> 12) & 0xff) << MPIDR_LEVEL_SHIFT(2);
 
-#ifdef CONFIG_ARM64_HISI_IPIV
-	if (static_branch_unlikely(&ipiv_enable)) {
-		/*
-		 * To avoid sending multi-SGIs in guest OS, make aff1/aff2 unique
-		 */
-		mpidr = (vcpu->vcpu_id & 0x0f) << MPIDR_LEVEL_SHIFT(1);
-		mpidr |= ((vcpu->vcpu_id >> 4) & 0xff) << MPIDR_LEVEL_SHIFT(2);
-	}
-#endif
-
 	mpidr |= (1ULL << 31);
 	vcpu_write_sys_reg(vcpu, mpidr, MPIDR_EL1);
 
diff --git a/drivers/irqchip/irq-gic-v3.c b/drivers/irqchip/irq-gic-v3.c
index 59354bd3e4be..06ca53cdf796 100644
--- a/drivers/irqchip/irq-gic-v3.c
+++ b/drivers/irqchip/irq-gic-v3.c
@@ -106,8 +106,12 @@ DEFINE_STATIC_KEY_FALSE(gic_nonsecure_priorities);
 EXPORT_SYMBOL(gic_nonsecure_priorities);
 
 #ifdef CONFIG_ARM64_HISI_IPIV
+/* indicate if host supports IPIv */
 DEFINE_STATIC_KEY_FALSE(ipiv_enable);
 EXPORT_SYMBOL(ipiv_enable);
+
+/* indicate if guest is using IPIv */
+static bool hisi_pv_sgi_enabled;
 #endif
 
 /*
@@ -1386,8 +1390,8 @@ void gic_dist_enable_ipiv(void)
 	static_branch_enable(&ipiv_enable);
 
 	val = (0 << GICD_IPIV_CTRL_AFF_DIRECT_VPEID_SHIFT) |
-		(0 << GICD_IPIV_CTRL_AFF1_LEFT_SHIFT_SHIFT) |
-		(4 << GICD_IPIV_CTRL_AFF2_LEFT_SHIFT_SHIFT) |
+		(4 << GICD_IPIV_CTRL_AFF1_LEFT_SHIFT_SHIFT) |
+		(12 << GICD_IPIV_CTRL_AFF2_LEFT_SHIFT_SHIFT) |
 		(7 << GICD_IPIV_CTRL_VM_TABLE_INNERCACHE_SHIFT) |
 		(2 << GICD_IPIV_CTRL_VM_TABLE_SHAREABILITY_SHIFT);
 	writel_relaxed(val, gic_data.dist_base + GICD_IPIV_CTRL);
@@ -1506,7 +1510,15 @@ static void gic_ipi_send_mask(struct irq_data *d, const struct cpumask *mask)
 		u64 cluster_id = MPIDR_TO_SGI_CLUSTER_ID(gic_cpu_to_affinity(cpu));
 		u16 tlist;
 
+#ifdef CONFIG_ARM64_HISI_IPIV
+		if (!hisi_pv_sgi_enabled)
+			tlist = gic_compute_target_list(&cpu, mask, cluster_id);
+		else
+			tlist = 1 << (gic_cpu_to_affinity(cpu) & 0xf);
+#else
 		tlist = gic_compute_target_list(&cpu, mask, cluster_id);
+#endif
+
 		gic_send_sgi(cluster_id, tlist, d->hwirq);
 	}
 
@@ -2740,6 +2752,28 @@ static struct fwnode_handle *gic_v3_get_gsi_domain_id(u32 gsi)
 	return gsi_domain_handle;
 }
 
+#ifdef CONFIG_ARM64_HISI_IPIV
+static void hisi_pv_sgi_init(void)
+{
+	struct arm_smccc_res res;
+
+	arm_smccc_1_1_invoke(ARM_SMCCC_VENDOR_PV_SGI_FEATURES, &res);
+	if (res.a0 != SMCCC_RET_SUCCESS) {
+		pr_info("Not Support HiSilicon PV SGI!\n");
+		return;
+	}
+
+	arm_smccc_1_1_invoke(ARM_SMCCC_VENDOR_PV_SGI_ENABLE, &res);
+	if (res.a0 != SMCCC_RET_SUCCESS) {
+		pr_info("Disable HiSilicon PV SGI!\n");
+		return;
+	}
+
+	hisi_pv_sgi_enabled = true;
+	pr_info("Enable HiSilicon PV SGI!\n");
+}
+#endif
+
 static int __init
 gic_acpi_init(union acpi_subtable_headers *header, const unsigned long end)
 {
@@ -2792,6 +2826,10 @@ gic_acpi_init(union acpi_subtable_headers *header, const unsigned long end)
 	if (static_branch_likely(&supports_deactivate_key))
 		gic_acpi_setup_kvm_info();
 
+#ifdef CONFIG_ARM64_HISI_IPIV
+	hisi_pv_sgi_init();
+#endif
+
 	return 0;
 
 out_fwhandle_free:
-- 
2.43.0

